crates/precompiles/src/nonce/mod.rs

```diff
@@ -130,244 +124,252 @@ impl<'a, S: PrecompileStorageProvider> NonceManager<'a, S> {
         let mut storage = HashMapStorageProvider::new(1).with_spec(TempoHardfork::Allegretto);
-        let account = address!("0x1111111111111111111111111111111111111111");
-        let nonce_key = U256::from(5);
+        StorageContext::enter(&mut storage, || {
+            let account = address!("0x1111111111111111111111111111111111111111");
+            let nonce_key = U256::from(5);
 
-        {
-            let mut nonce_mgr = NonceManager::new(&mut storage);
-            nonce_mgr.increment_nonce(account, nonce_key).unwrap();
-        }
+            // First increment emits ActiveKeyCountChanged + NonceIncremented
+            let mut mgr = NonceManager::new();
+            mgr.increment_nonce(account, nonce_key)?;
 
-        let events = storage
-            .events
-            .get(&NONCE_PRECOMPILE_ADDRESS)
-            .expect("Could not get events");
-        assert_eq!(events.len(), 2,);
-
-        let expected_nonce_event = NonceEvent::NonceIncremented(INonce::NonceIncremented {
-            account,
-            nonceKey: nonce_key,
-            newNonce: 1,
-        });
-        assert_eq!(events[1], expected_nonce_event.into_log_data());
-
-        {
-            let mut nonce_mgr = NonceManager::new(&mut storage);
-            nonce_mgr.increment_nonce(account, nonce_key).unwrap();
-        }
-        let events = storage.events.get(&NONCE_PRECOMPILE_ADDRESS).unwrap();
-        assert_eq!(events.len(), 3);
-
-        let expected_nonce_event2 = NonceEvent::NonceIncremented(INonce::NonceIncremented {
-            account,
-            nonceKey: nonce_key,
-            newNonce: 2,
-        });
-        assert_eq!(events[2], expected_nonce_event2.into_log_data());
+            mgr.assert_emitted_events(vec![
+                NonceEvent::ActiveKeyCountChanged(INonce::ActiveKeyCountChanged {
+                    account,
+                    newCount: U256::ONE,
+                }),
+                NonceEvent::NonceIncremented(INonce::NonceIncremented {
+                    account,
+                    nonceKey: nonce_key,
+                    newNonce: 1,
+                }),
+            ]);
+
+            // Second increment on same key only emits NonceIncremented (no new key)
+            mgr.increment_nonce(account, nonce_key)?;
+            assert_eq!(mgr.emitted_events().len(), 3);
+
+            mgr.assert_emitted_events(vec![
+                NonceEvent::ActiveKeyCountChanged(INonce::ActiveKeyCountChanged {
+                    account,
+                    newCount: U256::ONE,
+                }),
+                NonceEvent::NonceIncremented(INonce::NonceIncremented {
+                    account,
+                    nonceKey: nonce_key,
+                    newNonce: 1,
+                }),
+                NonceEvent::NonceIncremented(INonce::NonceIncremented {
+                    account,
+                    nonceKey: nonce_key,
+                    newNonce: 2,
```

Comment:
nit: looks like we are asserting events twice when we could just be asserting all events at the end of the last call. This is the case in a few of the tests.